crates/dbsp/src/algebra/decimal.rs

```diff
@@ -0,0 +1,607 @@
+        out.inner = self.inner.rem(rhs.inner);
+
+        out
+    }
+}
+
+impl Num for SQLDecimal {
+    type FromStrRadixErr = String;
+
+    fn from_str_radix(str: &str, radix: u32) -> Result<Self, Self::FromStrRadixErr> {
+        let inner = Decimal::from_str_radix(str, radix).map_err(|e| e.to_string())?;
+
+        Self::try_from(inner)
+    }
+}
+
+impl Signed for SQLDecimal {
+    fn abs(&self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.abs();
+        out.apply_scaling_rules();
+        out
+    }
+
+    fn abs_sub(&self, other: &Self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.abs_sub(&other.inner);
+        out.apply_scaling_rules();
+        out
+    }
+
+    fn signum(&self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.signum();
+        out.apply_scaling_rules();
+        out
+    }
+
+    fn is_positive(&self) -> bool {
+        self.inner.is_sign_positive()
+    }
+
+    fn is_negative(&self) -> bool {
+        self.inner.is_sign_negative()
+    }
+}
+
+impl ToPrimitive for SQLDecimal {
+    fn to_i64(&self) -> Option<i64> {
+        self.inner.to_i64()
+    }
+
+    fn to_u64(&self) -> Option<u64> {
+        self.inner.to_u64()
+    }
+
+    fn to_f32(&self) -> Option<f32> {
+        self.inner.to_f32()
+    }
+
+    fn to_f64(&self) -> Option<f64> {
+        self.inner.to_f64()
+    }
+}
+
+impl FromPrimitive for SQLDecimal {
+    fn from_i64(n: i64) -> Option<Self> {
+        let inner = Decimal::from_i64(n)?;
+
+        Self::try_from(inner).ok()
+    }
+
+    fn from_u64(n: u64) -> Option<Self> {
+        let inner = Decimal::from_u64(n)?;
+
+        Self::try_from(inner).ok()
+    }
+
+    fn from_f32(n: f32) -> Option<Self> {
+        let inner = Decimal::from_f32(n)?;
+
+        Self::try_from(inner).ok()
+    }
+
+    fn from_f64(n: f64) -> Option<Self> {
+        let inner = Decimal::from_f64(n)?;
+
+        Self::try_from(inner).ok()
+    }
+}
+
+impl Display for SQLDecimal {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.inner)
+    }
+}
+
+impl From<i32> for SQLDecimal {
+    fn from(value: i32) -> Self {
+        let inner = Decimal::from(value);
+
+        Self::try_from(inner).unwrap()
+    }
+}
+
+impl From<i64> for SQLDecimal {
+    fn from(value: i64) -> Self {
+        let inner = Decimal::from(value);
+        let scale = inner.scale();
+
+        SQLDecimal {
+            inner,
+            max_precision: Precision::max(),
+            max_scale: Scale::try_from(scale).unwrap(),
+        }
+    }
+}
+
+impl From<isize> for SQLDecimal {
+    fn from(value: isize) -> Self {
+        let inner = Decimal::from(value);
+        let scale = inner.scale();
+
+        SQLDecimal {
+            inner,
+            max_precision: Precision::max(),
+            max_scale: Scale::try_from(scale).unwrap(),
+        }
+    }
+}
+
+impl Zero for SQLDecimal {
+    fn zero() -> Self {
```

Comment:
I am not sure we can have this, we always need a precision and scale.

Reply:
We need to implement the `Zero` trait to implement the `Num` trait.